File: backend/app/schemas/audit.py
from enum import Enum
from pydantic import BaseModel, Field, HttpUrl
from typing import List, Optional, Dict
from datetime import datetime
from app.schemas.evidence import BrowserEvidence, ViewportAuditResult
from app.schemas.issue import Issue


class AuditStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class CreateAuditRequest(BaseModel):
    url: str = Field(..., description="Target web application URL to audit")
    viewports: List[str] = Field(
        default=["desktop", "mobile"],
        description="Viewport sizes to test: desktop (1440x900), mobile (390x844)"
    )
    baseline_audit_id: Optional[str] = Field(
        None,
        description="Optional baseline audit ID for before/after fix verification comparison"
    )


class AuditSummaryStats(BaseModel):
    total_issues: int = 0
    critical_count: int = 0
    high_count: int = 0
    medium_count: int = 0
    low_count: int = 0
    info_count: int = 0
    console_error_count: int = 0
    network_failure_count: int = 0
    layout_issue_count: int = 0
    accessibility_issue_count: int = 0


class AuditResult(BaseModel):
    audit_id: str = Field(..., description="Unique UUID for this audit run")
    target_url: Optional[str] = Field(None, description="Target web application URL")
    url: str
    status: AuditStatus
    created_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    desktop: Optional[ViewportAuditResult] = None
    mobile: Optional[ViewportAuditResult] = None
    evidence: Optional[BrowserEvidence] = None
    issues: List[Issue] = []
    findings: List[Issue] = []
    stats: AuditSummaryStats = Field(default_factory=AuditSummaryStats)
    error_message: Optional[str] = None



class AuditComparison(BaseModel):
    baseline_audit_id: str
    new_audit_id: str
    created_at: datetime
    fixed_issues: List[Issue] = Field(default=[], description="Issues resolved in new audit")
    remaining_issues: List[Issue] = Field(default=[], description="Issues still present in new audit")
    new_issues: List[Issue] = Field(default=[], description="New issues introduced in new audit")
    regressions: List[Issue] = Field(default=[], description="Reintroduced or worsened issues")


class DeveloperFixPrompt(BaseModel):
    audit_id: str
    target_issues: List[Issue]
    fix_prompt: str = Field(..., description="Context-rich prompt formatted for Antigravity, Cursor, Claude, or VS Code")
    suggested_files: List[str] = []


class RetestAuditResponse(BaseModel):
    retest_audit: AuditResult = Field(..., description="The newly executed retest audit result")
    comparison: AuditComparison = Field(..., description="Deterministic comparison between baseline and retest audit")

File: frontend/src/types/audit.ts
export type IssueSeverity = 'critical' | 'high' | 'medium' | 'low' | 'info';

export type IssueCategory =
  | 'layout'
  | 'responsive'
  | 'console_error'
  | 'console'
  | 'network_failure'
  | 'broken_resource'
  | 'seo'
  | 'accessibility'
  | 'performance'
  | 'interaction';

export interface Issue {
  issue_id: string;
  category: IssueCategory;
  severity: IssueSeverity;
  title: string;
  description: string;
  selector?: string;
  viewport?: string;
  evidence_references?: string[];
  root_cause_analysis?: string;
  recommended_fix?: string;
}

export interface BrowserViewport {
  name: string;
  width: number;
  height: number;
  device_scale_factor: number;
}

export interface ConsoleLogEntry {
  timestamp: string;
  level: string;
  text: string;
  location?: string;
}

export interface NetworkFailure {
  timestamp: string;
  url: string;
  method: string;
  status_code?: number;
  error_text: string;
}

export interface ResponsiveMetrics {
  viewport_width: number;
  viewport_height: number;
  document_scroll_width: number;
  document_client_width: number;
  horizontal_overflow: number;
}

export interface PageMetadata {
  initial_url: string;
  final_url: string;
  title: string;
  http_status?: number;
  page_load_success: boolean;
  error_message?: string;
}

export interface ViewportAuditResult {
  viewport: BrowserViewport;
  page: PageMetadata;
  screenshot_artifact_id?: string;
  screenshot_url?: string;
  console_errors: ConsoleLogEntry[];
  console_logs?: ConsoleLogEntry[];
  network_failures: NetworkFailure[];
  responsive: ResponsiveMetrics;
}

export interface LayoutIssue {
  viewport: BrowserViewport;
  selector: string;
  issue_type: string;
  description: string;
}

export interface AccessibilityIssue {
  rule_id: string;
  impact: string;
  selector: string;
  description: string;
}

export interface BrowserEvidence {
  url: string;
  timestamp: string;
  desktop?: ViewportAuditResult;
  mobile?: ViewportAuditResult;
  viewports_tested: BrowserViewport[];
  screenshot_paths: string[];
  console_errors: ConsoleLogEntry[];
  network_failures: NetworkFailure[];
  layout_issues: LayoutIssue[];
  accessibility_issues: AccessibilityIssue[];
}

export interface AuditSummaryStats {
  total_issues: number;
  critical_count: number;
  high_count: number;
  medium_count: number;
  low_count: number;
  info_count: number;
  console_error_count: number;
  network_failure_count: number;
  layout_issue_count: number;
  accessibility_issue_count: number;
}

export type AuditStatus = 'pending' | 'running' | 'completed' | 'failed';

export interface AuditResult {
  audit_id: string;
  target_url?: string;
  url: string;
  status: AuditStatus;
  created_at: string;
  started_at?: string;
  completed_at?: string;
  desktop?: ViewportAuditResult;
  mobile?: ViewportAuditResult;
  evidence?: BrowserEvidence;
  issues: Issue[];
  findings?: Issue[];
  stats: AuditSummaryStats;
  error_message?: string;
}

export interface CreateAuditRequest {
  url: string;
  viewports?: string[];
  baseline_audit_id?: string;
}

export interface AuditComparison {
  baseline_audit_id: string;
  new_audit_id: string;
  created_at: string;
  fixed_issues: Issue[];
  remaining_issues: Issue[];
  new_issues: Issue[];
  regressions: Issue[];
}

export interface DeveloperFixPrompt {
  audit_id: string;
  target_issues: Issue[];
  fix_prompt: string;
  suggested_files: string[];
}
